nixos-rebuild: Add `nixos-rebuild repl`

roberth opened the pull request:
###### Description of changes

Add `nixos-rebuild repl`, which loads a `nix repl` for you.

If you use flakes, brings the flake itself into scope as well.

Apologies to the non-flake users; your repl isn't quite as fancy, but at least evaluates your config exactly as you would expect, unlike flakes which are only evaluated impurely for now.

Tested by hand, as evaluating in the sandbox is tricky.

**Try it**

Flake in `~/config`

```console
nix run github:hercules-ci/nixpkgs/nixos-rebuild-repl#nixos-rebuild -- repl --flake ~/config
```

Non-flake

```console
nix --extra-experimental-features flakes\ nix-command run github:hercules-ci/nixpkgs/nixos-rebuild-repl#nixos-rebuild -- repl
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos-rebuild: Add nixos-rebuild repl

Apologies to the non-flake users; your repl isn't quite as fancy,
but at least evaluates your config exactly as you would expect,
unlike flakes which are only evaluated impurely for now.

SuperSandro2000 commented on nixos/doc/manual/installation/changing-config.chapter.md:
````diff
@@ -55,6 +55,14 @@ which causes the new configuration (and previous ones created using
+# nixos-rebuild repl
+```
+
+Your configuration is loaded into the `config` variable. Use tab for autocompletion, use the `:r` command to reload the configuration files. See `:?` or [`nix repl` in the Nix manual](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl.html) to learn more.
````
Either I do not understand how ``:r`` is supposed to work or it is broken. For me it just clears the loaded variables and loads --file again and totally ignored ``:lf``.

roberth replied:
`:lf` behaves different from `:l`, and taken together I consider it to be quite broken, yes.
- https://github.com/NixOS/nix/issues/8753

SuperSandro2000 reviewed (commented):
This feels like feature creep to me because it can be shortened to ``nix repl``, ``:lf /etc/nixos`` (or whatever path your config is) and ``nixosConfiguration.$HOSTNAME.``

roberth commented:
Right. Reminds me of "dropbox is just ftp". The difference is real. It actually solves the problem and makes it nice to use (especially the flakes case, as I left the non-flake case very basic), but you couldn't believe it.

doronbehar commented:
Hey, I like this! A few comments:

If no `--flake` argument is given, and in my case there is no `/etc/nixos` directory, and my system's flake is located in a non trivial path, I get into a repl, and no nice "message of the day" is presented. Would it be possible to fix that?

In [this issue discussion](https://github.com/NixOS/nixpkgs/issues/232679#issuecomment-1556761770), I wrote to myself and to others how to test someone else's configuration, and possibly with a specific nixpkgs branch - to test whether module changes etc... Would it be possible to add an argument to `nixos-rebuild repl` that will enable using a specific nixpkgs for instance? or a file such as `raito.nix` as mentioned in that comment?

roberth commented:
> If no `--flake` argument is given, and in my case there is no `/etc/nixos` directory, and my system's flake is located in a non trivial path, I get into a repl, and no nice "message of the day" is presented. Would it be possible to fix that?

It may be.
In non-flakes, the `nixos-rebuild` command and its expressions take input from `NIX_PATH` and such, so `-I nixos=config=~/nixos/foo.nix` should just work as usual.

As for the motd, it's quite specific to how the flakes-based repl works. Perhaps things could be refactored a bit, but I can't do it now. I don't think this should be considered a blocker.

roberth commented:
Added this

```
    # This is a very end user command, implemented using sub-optimal means.
    # You should feel free to improve its behavior, as well as resolve tech
    # debt in "breaking" ways. Humans adapt quite well.
```

I believe leaving this feature in an unmerged PR would be a waste (and the only alternative) and an obstacle to contributors who want to not only use, but also improve the feature. So I would suggest to merge this soon.

doronbehar reviewed (approved):
> I believe leaving this feature in an unmerged PR would be a waste (and the only alternative) and an obstacle to contributors who want to not only use, but also improve the feature. So I would suggest to merge this soon.

I agree. I feel like there is agreement that there is good value in pushing towards having this `repl` functionality. I haven't tested the new changes, but nothing seems totally new, besides the new comment near the implementation. Should be good to go after CI is green IMO.

roberth commented:
I've only touched up the formatting to feel a bit more Nix-like. Thanks for the approval!
